### PR TITLE
Avoid unnecessary Image operations

### PR DESCRIPTION
**Pull request by radarhere**
- Avoid unnecessarily loading an image twice
- In Image's convert method, if image is already in the destination mode, do not convert

### CODE REVIEW

File: PIL/Image.py [hunk 1]
```diff
@@ -1589,11 +1591,10 @@ def putpixel(self, xy, value):
         :param value: The pixel value.
         """
 
-        self.load()
         if self.readonly:
             self._copy()
             self.pyaccess = None
```
Comment (homm) [inline PIL/Image.py]: `self._copy()` already sets `pyaccess` to None

**Comment by homm** [comment]
```python
def _mutable(self):
    if self.readonly:
        self._copy()
    else:
        self.load()
```

**Comment by homm** [comment]
First of all, `im.load()` is always noop when called in second and subsequent times. It's very cheap call. So, it's nice to avoid the double calling for some operations, but it is not necessary. I would agree, that we can avoid calling nearby `self.readonly`, but as of `Image.convert()` method, it is too complicated. I'd prefer to have one call in the beginning. Something like this:

```python
def convert(...):
    self.load()

    if not mode and self.mode == "P":
        # determine default mode
        if self.palette:
            mode = self.palette.mode
        else:
            mode = "RGB"

    if not mode or (mode == self.mode and matrix is None):
        return self.copy()
        
...
```

File: PIL/Image.py (hunk 1 above)
Comment (radarhere) [reply PIL/Image.py]: I have created a separate PR for this, #2714

**Comment by radarhere** [comment]
Okay, I've updated the commits. See what you think.

File: PIL/Image.py [hunk 2]
```diff
@@ -592,6 +592,12 @@ def _copy(self):
         self.pyaccess = None
         self.readonly = 0
 
+    def _mutable(self):
+        if self.readonly:
```
Comment (wiredfool) [inline PIL/Image.py]: I don't like the name of this, as it shows in the calling code. The bare call to `self._mutable` while clearly having a side effect, doesn't clearly explain what's going on. 

I'd be happier with something like `_ensure_mutable`.

**Comment by wiredfool** [comment]
merged in #2767

**Pull request merged**